ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableFlatEvents.scala

```diff
@@ -62,47 +63,58 @@ private[events] trait EventsTableFlatEvents { this: EventsTable =>
 
   val flatEventParser: RowParser[Entry[Event]] = createdFlatEventParser | archivedFlatEventParser
 
+  private val selectColumns =
+    Seq(
+      "event_offset",
+      "transaction_id",
+      "ledger_effective_time",
+      "case when submitter = {requesting_parties} then command_id else '' end as command_id",
+      "workflow_id",
+      "participant_events.event_id",
+      "contract_id",
+      "template_package_id",
+      "template_name",
+      "create_argument",
+      "create_signatories",
+      "create_observers",
+      "create_agreement_text",
+      "create_key_value",
+      "array_agg(event_witness) as event_witnesses",
+    ).mkString(", ")
+
+  private val flatEventsTable =
+    "participant_events natural join participant_event_flat_transaction_witnesses"
+
+  private val groupByColumns =
+    Seq(
+      "event_offset",
+      "transaction_id",
+      "ledger_effective_time",
+      "command_id",
+      "workflow_id",
+      "participant_events.event_id",
+      "contract_id",
+      "template_package_id",
+      "template_name",
+      "create_argument",
+      "create_signatories",
+      "create_observers",
+      "create_agreement_text",
+      "create_key_value",
+    ).mkString(", ")
+
+  private val lookupFlatTransactionByIdQuery: SqlQuery =
+    SQL(
+      s"select $selectColumns from $flatEventsTable where transaction_id = {transaction_id} and event_witness in ({requesting_parties}) group by ($groupByColumns)"
+    )
```

Comment:
For safety around the parameters, you could also do something like:

```suggestion
  private def lookupFlatTransactionByIdQuery(
      transactionId: TransactionId,
      requestingParties: Set[Party],
  ): SimpleSql[Row] =
    SQL"select #$selectColumns from #$flatEventsTable where transaction_id = $transactionId and event_witness in ($requestingParties) group by (#$groupByColumns)"
```

Your call though; I don't mind either way.

Reply:
It's very nice to know about the syntax, can it still take advantage of `PreparedStatement` caching?

Reply:
Yup. [From the docs:](http://playframework.github.io/anorm/#string-interpolation)

> By using `#$value` instead of `$value`, interpolated value will be part of the prepared statement, rather being passed as a parameter when executing this SQL statement.

Reply:
This is very cool, thanks for sharing.

Reply:
I added [this](https://github.com/digital-asset/daml/pull/5241/commits/926c17135222937bb5eb0a984589ae04714cfe60) to make it work. Unfortunately I had to move one columns out of the shared ones because it needs a parameter. Let me know if you have better ways to do it.